bpfd/src/multiprog/tc.rs

```diff
@@ -0,0 +1,243 @@
+                let path = format!("{base}/dispatcher_{if_index}_{}/link_{k}", self.revision);
+                new_link.pin(path).map_err(|_| BpfdError::UnableToPin)?;
+            } else {
+                let mut bpf = BpfLoader::new()
+                    .map_pin_path(format!("{RTDIR_FS_MAPS}/{k}"))
+                    .extension(&v.data.section_name)
+                    .load_file(&v.data.path)
+                    .map_err(BpfdError::BpfLoadError)?;
+
+                let ext: &mut Extension = bpf
+                    .program_mut(&v.info.metadata.name)
+                    .ok_or_else(|| BpfdError::SectionNameNotValid(v.info.metadata.name.clone()))?
+                    .try_into()?;
+
+                let target_fn = format!("prog{i}");
+
+                ext.load(dispatcher.fd().unwrap(), &target_fn)?;
+                ext.pin(format!("{RTDIR_FS}/prog_{k}"))
+                    .map_err(|_| BpfdError::UnableToPin)?;
+                let new_link_id = ext.attach()?;
+                let new_link = ext.take_link(new_link_id)?;
+                let fd_link: FdLink = new_link.into();
+                let base = match self.direction {
+                    Direction::Ingress => RTDIR_FS_TC_INGRESS,
+                    Direction::Egress => RTDIR_FS_TC_EGRESS,
+                };
+                fd_link
+                    .pin(format!(
+                        "{base}/dispatcher_{if_index}_{}/link_{k}",
+                        self.revision,
+                    ))
+                    .map_err(|_| BpfdError::UnableToPin)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn save(&self) -> Result<(), BpfdError> {
+        let base = match self.direction {
+            Direction::Ingress => RTDIR_TC_INGRESS_DISPATCHER,
+            Direction::Egress => RTDIR_TC_EGRESS_DISPATCHER,
+        };
+        let path = format!("{base}/{}-{}", self.if_index, self.revision);
+        serde_json::to_writer(&fs::File::create(path).unwrap(), &self)
+            .map_err(|e| BpfdError::Error(format!("can't save state: {}", e)))?;
+        Ok(())
+    }
+
+    pub(crate) fn delete(&self, _full: bool) -> Result<(), BpfdError> {
+        let base = match self.direction {
+            Direction::Ingress => RTDIR_TC_INGRESS_DISPATCHER,
+            Direction::Egress => RTDIR_TC_EGRESS_DISPATCHER,
+        };
+        let path = format!("{base}/{}-{}", self.if_index, self.revision);
+        fs::remove_file(path)
+            .map_err(|e| BpfdError::Error(format!("unable to cleanup state: {}", e)))?;
+
+        let base = match self.direction {
+            Direction::Ingress => RTDIR_FS_TC_INGRESS,
+            Direction::Egress => RTDIR_FS_TC_EGRESS,
+        };
+        let path = format!("{base}/dispatcher_{}_{}", self.if_index, self.revision);
+        fs::remove_dir_all(path)
+            .map_err(|e| BpfdError::Error(format!("unable to cleanup state: {}", e)))?;
+        // Dispatcher will be detached when this object is dropped
```

Comment:
It's actually not dropped presumably because it's not "detach on drop" in Aya.

Reply:
Ack. Updated the comment and added a FIXME